### PR TITLE
Stack Layout Optimizations

### PR DESCRIPTION
**Pull request by kyleve**
This PR optimizes calculating the axis measurements in `Stack` elements (`Row` and `Column`) when there are N fixed and 1 flexible elements. This generally gets us a ~50% speedup in many cases where this sort of layout is prevalent (eg, a `MarketRow` with a leading accessory, one label, and then trailing content).

**Before (running `test_rows_with_one_flexible_element`):**
`Iterations: 6, Average Time: 0.3376373251279195`

**After:**
`Iterations: 12, Average Time: 0.17800241708755493`

This is done by optimizing the axis measurement algorithm to take into account the fixed-ness and flexible-ness of content in the stack. Before this change, we would (note: some simplifying done here)...

1. Measure every child element in the full axis size of the stack to determine their size.
2. Find the fixed width of the fixed width items in the stack, use this to re-measure the flexible elements (paraphrased).
3. Re-measure the flexible elements in the full width minus the fixed width items.

Notable, step 3 **always caused a second measurement pass on the flexible elements**. This got worse the deeper the element tree downstream from that flexible element, since 2 measurements could become 4, could become 8, etc...

To resolve, I changed the axis segment calculation to...

1. Find the fixed size elements.
2. Measure them in the full size.
3. Add up the total axis size of the fixed size elements.
4. Use the full axis size minus the axis size of fixed elements to measure the flexible element(s). 

In the cases that there is only 1 flexible element (quite common), this results in one less measurement pass on the flexible element, which can dramatically improve rendering times. This is because fewer measurements overall occur. Eg, in `test_rows_with_one_flexible_element_simple`, with 10 child elements, before this change, the labels are measured a total of 80 times. After it, they are measured only 30 times.

#### `CacheTree` rounding
Moved to a follow up PR: https://github.com/square/Blueprint/pull/362 ~Due to float addition and subtraction that happens throughout layout, it was common to end up with `210.0` getting turned into `209.999997` (etc) throughout. This was causing needless cache misses. To resolve, I introduced pixel-based rounding to the cache tree so the cached constraints land on a display pixel. I'm not 100% sure this is the best approach, but it's a starting point.~

#### Further optimization
There's some further optimization we could do here, since a lot the remaining measurements are semantically identical (with varying size constraints that all result in the same outcome), but optimizing this is a lot harder. We can also do some optimization of the cross segment calculation, but I will follow up with this one.

### CODE REVIEW

File: BUILD.bazel
Comment (kyleve) [inline BUILD.bazel:1]: Delete?

File: BlueprintUI/Sources/BlueprintView/BlueprintView.swift
Comment (kyleve) [inline BlueprintUI/Sources/BlueprintView/BlueprintView.swift:211]: TODO: Is this the right fallback?

File: BlueprintUI/Sources/Internal/CacheTree.swift
Comment (kyleve) [inline BlueprintUI/Sources/Internal/CacheTree.swift:34]: I think instead of up, we want to closest value.

File: BlueprintUI/Tests/StackTests.swift
Comment (kyleve) [inline BlueprintUI/Tests/StackTests.swift:1186]: Finish me

File: BlueprintUICommonControls/Sources/AttributedLabel.swift
Comment (kyleve) [inline BlueprintUICommonControls/Sources/AttributedLabel.swift:96]: Revert before merge

File: BlueprintUICommonControls/Tests/Sources/PerformancePlayground.swift
Comment (kyleve) [inline BlueprintUICommonControls/Tests/Sources/PerformancePlayground.swift:27]: Revert before merge

File: Package.swift
Comment (kyleve) [inline Package.swift:24]: Revert before merge

File: BlueprintUI/Sources/Internal/CacheTree.swift
Comment (kyleve) [inline BlueprintUI/Sources/Internal/CacheTree.swift:37]: This breaks one test; which is interesting; I _think_ because test elements are relying on their measurement constraint to be unrounded (even unrounded off of screen pixel values, eg 0.333). This feels "wrong", but didn't change it.

File: BlueprintUI/Sources/Layout/Stack.swift
Comment (kyleve) [inline BlueprintUI/Sources/Layout/Stack.swift:276]: Remove before merging.

File: BlueprintUI/Sources/Layout/Stack.swift
Comment (kyleve) [inline BlueprintUI/Sources/Layout/Stack.swift:504]: Looking for input on what to do here – eg if there's one or more fixed elements that take up all the space. I was thinking we may just want to skip measuring the flexible elements entirely, since they basically can't lay out.

File: BlueprintUI/Tests/StackTests.swift
Comment (kyleve) [inline BlueprintUI/Tests/StackTests.swift:1177]: These changes b/c the flexible elements are now measured in the _available_ space, not _all_ the space.

File: BlueprintUI/Sources/Layout/Stack.swift
Comment (n8chur) [reply BlueprintUI/Sources/Layout/Stack.swift:508]: I might be misunderstanding the issue, but I'd expect we could skip measurement of those flexible items and just use a size of `0`, right?

**Review by n8chur — COMMENTED** [review]
Would love to get a review on this from @watt—these look like reasonable changes on the surface but having someone with more experience here would be really helpful.

File: BlueprintUI/Sources/Layout/Stack.swift
Comment (kyleve) [reply BlueprintUI/Sources/Layout/Stack.swift:508]: I think we could – let me see what happened before at layout time.

File: BlueprintUI/Sources/Internal/CacheTree.swift
Comment (kyleve) [inline BlueprintUI/Sources/Internal/CacheTree.swift:25]: - [ ] Move `CacheTree` changes to their own PR

**Comment by kyleve** [comment]
Moved CacheTree changes to a new PR: https://github.com/square/Blueprint/pull/362

File: BlueprintUI/Sources/Layout/Stack.swift
Comment (kyleve) [reply BlueprintUI/Sources/Layout/Stack.swift]: I added a test (`test_fixed_elements_overflow_behavior`), and the behavior here mirrors the old behavior – which seems wrong; but at least it's consistent. I'll follow up with lower bounding this to zero.

Comment (kyleve) [reply BlueprintUI/Sources/Layout/Stack.swift]: https://github.com/square/Blueprint/pull/363

**Comment by kyleve** [comment]
Ran the Register snapshot tests over this again, and there's a bunch of tiny lil 1px rounding changes, but I only found a couple meaningful differences, which seem to be due to incorrect configuration of the Rows/Columns in the first place, and will be easy to fix!

File: BlueprintUI/Tests/StackTests.swift
Comment (n8chur) [inline BlueprintUI/Tests/StackTests.swift:1296]: what are the counts for each of these before you made these changes?

**Review by n8chur — APPROVED** [review]
Looks reasonable to me but would really like to get a review from @watt before this lands if possible!

File: BlueprintUI/Tests/StackTests.swift
Comment (kyleve) [reply BlueprintUI/Tests/StackTests.swift:1296]: Some were 4, and some were 2.

File: BlueprintUI/Sources/Layout/Stack.swift
Comment (watt) [inline BlueprintUI/Sources/Layout/Stack.swift:525]: Ahh OK, I was wondering how you reduced measurement counts entirely within `_axisSegments`, but it seems to be the same number of calls. If I'm understanding right, it looks like this optimization really hinges on that `CacheTree` change in order to shortcut measurements from `_crossSegments` before they happen. The change to flexible child constraints makes the input values match for the 1-flexible-child case.

**Review by watt — COMMENTED** [review]
Seeing how this works by relying on pre-caching measurements, rather than explicitly skipping measurements within the stack layout itself, I'm a bit nervous about this change. Gotta think about it a bit more 🤔

File: BlueprintUI/Sources/Internal/CacheTree.swift
Comment (kyleve) [reply BlueprintUI/Sources/Internal/CacheTree.swift:38]: Yeahhhhh, I was thinking about this too. It felt rare / weird enough that the tradeoff was worth it, but your point is valid – we can't guarantee folks _don't_ rely on this, so I am also fine just removing it... It _does_ cut the measurement cost in half once again though... Trying to think about how we could keep the "good" parts here...

File: BlueprintUI/Tests/StackTests.swift
Comment (kyleve) [reply BlueprintUI/Tests/StackTests.swift:1180]: Agree. My view on current system:

(a) I do not like the current system, and dare I say no one (outside of like 3 people, and 2 of us are on this PR) understand it (I do not like green eggs and ham)
(b) Every time I see someone trying to use a non-binary priority (eg 0 or 1), it's used wrong
(c) I think this change actually makes things more understandable / behave as you expect for the common case – eg because we're now measuring flexible elements in the actual size they have, it fixes many of the issues called out in https://github.com/square/Blueprint/pull/265 – which you can see from running this in Market – the weird "huh, that row layout sure is wrong" bugs are fixed.

I will update the tests here, but I'd also like to follow up with turning growing and shrinking into bools vs ints, since, yeah...

File: BlueprintUI/Sources/Layout/Stack.swift
Comment (kyleve) [reply BlueprintUI/Sources/Layout/Stack.swift:540]: - [ ] Bring back optimization for unconstrained cases

File: BlueprintUI/Sources/Internal/CacheTree.swift
Comment (kyleve) [reply BlueprintUI/Sources/Internal/CacheTree.swift:38]: (I was also trying to figure out how to lift this level of caching up to this type only, but not really a doable thing rn..)

Comment (kyleve) [reply BlueprintUI/Sources/Internal/CacheTree.swift:38]: Oh I have an idea

Comment (kyleve) [reply BlueprintUI/Sources/Internal/CacheTree.swift:38]: Hmm, thinking about this more... This is OK? Eg you don't have to respect the size returned here. And if you still stretch the returned size, the aforementioned layout will still work... I think?

(Ok we've returned to where I was when I first wrote this code and had the same thought. Trying to figure out how to prove this case beyond "the snapshots in SPOS still pass".)

Comment (watt) [reply BlueprintUI/Sources/Internal/CacheTree.swift:38]: After sleeping on it a bit, I guess I'm OK with this.

I do think that lifting this logic up to to be done explicitly in stack layout is still probably the more "right" thing, for a couple reasons:
1. the previously mentioned correctness argument, since you could definitely build an element that this breaks
2. the way that this interacts with stack layout is not very obvious. If someone else tweaked stack layout a bit we could lose this optimization without knowing why. Or tests would fail for frustratingly unclear reasons.

But practically speaking, I can only come up with hypothetical cases that cause (1), and I'm less worried about (2) knowing that a layout rewrite is likely right around the corner to support SPL. Worth the tradeoffs.

**Review by watt — APPROVED** [review]

File: BlueprintUI/Sources/Internal/CacheTree.swift
Comment (kyleve) [reply BlueprintUI/Sources/Internal/CacheTree.swift:38]: I'm gonna hold off on merging now, and I would like to talk through this a bit more next week – I think these are very valid points and wanna make sure I don't paint us into a corner with this change. I also had a couple other ideas for how to lift / manage the caching a bit better for _only_ the stack layout which I need to play around with.

File: BlueprintUI/Sources/Layout/Stack.swift
Comment (kyleve) [reply BlueprintUI/Sources/Layout/Stack.swift:525]: Sorry, just saw this comment as I come back to this PR. The cache tree changes get us from 2 measurements to 1 – if you run the `test_measurementCounts` test and turn off `BlueprintStackUpdates.isOptimizedLayoutEnabled`, you can also see the main optimization, of some measurements happening 3, 4, times, etc.

File: BlueprintUI/Sources/Internal/CacheTree.swift
Comment (kyleve) [reply BlueprintUI/Sources/Internal/CacheTree.swift:38]: I think I have a way to do this on the _get_ end, not the set end. Testing to see what happens.

Comment (kyleve) [reply BlueprintUI/Sources/Internal/CacheTree.swift:38]: Actually wait, that won't work unless we have the concept of cached measurements and if they're natural sizes or not. I might just remove this for now, gonna play with it a bit more.

Comment (kyleve) [reply BlueprintUI/Sources/Internal/CacheTree.swift:38]: Yeah, gonna just get rid of this for now. It reduces the benefit a bit, but it still cuts many measurements down to 2 passes instead of 2 * depth.

Comment (kyleve) [reply BlueprintUI/Sources/Internal/CacheTree.swift:38]: Eg, with the flag disabled: 
image

File: BlueprintUI/Sources/Layout/Stack.swift
Comment (kyleve) [reply BlueprintUI/Sources/Layout/Stack.swift:540]: Done

File: BlueprintUI/Sources/Layout/Stack.swift
Comment (kyleve) [reply BlueprintUI/Sources/Layout/Stack.swift]: Removed

**Pull request merged**
